core/src/main/java/io/questdb/cutlass/http/processors/PrometheusMetricsProcessor.java

```diff
@@ -24,42 +24,151 @@
     private final Scrapable metrics;
+    private final RequestStatePool pool;
     private final boolean requiresAuthentication;
 
-    public PrometheusMetricsProcessor(Scrapable metrics, HttpMinServerConfiguration configuration) {
+    public PrometheusMetricsProcessor(Scrapable metrics, HttpMinServerConfiguration configuration, RequestStatePool pool) {
         this.metrics = metrics;
         this.requiresAuthentication = configuration.isHealthCheckAuthenticationRequired();
+        this.pool = pool;
     }
 
     @Override
     public void onRequestComplete(HttpConnectionContext context) throws PeerDisconnectedException, PeerIsSlowToReadException {
-        HttpChunkedResponseSocket r = context.getChunkedResponseSocket();
+        final RequestState state = setupState(context);
+
+        // We double-buffer the metrics response.
+        // This is because we send back chunked responses and each chunk may not be large enough.
+        metrics.scrapeIntoPrometheus(state.sink);
+
+        final HttpChunkedResponseSocket r = context.getChunkedResponseSocket();
         r.status(200, CONTENT_TYPE_TEXT);
         r.sendHeader();
+        sendResponse(r, state);
+    }
 
-        metrics.scrapeIntoPrometheus(r);
-
-        r.done();
+    @Override
+    public boolean requiresAuthentication() {
+        return requiresAuthentication;
     }
 
+    /**
+     * Continues after `PeerIsSlowToReadException` was thrown
+     * by `onRequestComplete` or earlier call to `resumeSend`.
+     */
     @Override
     public void resumeSend(HttpConnectionContext context) throws PeerDisconnectedException, PeerIsSlowToReadException {
+        // Send the remainder of the current, partially sent, chunk.
         context.resumeResponseSend();
+
+        // Send any remaining chunks, if any.
+        final RequestState state = LV.get(context);
+        assert state != null;
+        final HttpChunkedResponseSocket r = context.getChunkedResponseSocket();
+        sendResponse(r, state);
     }
 
-    @Override
-    public boolean requiresAuthentication() {
-        return requiresAuthentication;
+    /** Send the next chunk and return true if no more chunks need to be sent. */
+    private boolean sendNextChunk(HttpChunkedResponseSocket r, RequestState state) throws PeerIsSlowToReadException, PeerDisconnectedException {
+        final int pending = state.sink.length() - state.written;
+        final int wrote = r.writeBytes(state.sink.ptr() + state.written, pending);
+        state.written += wrote;
+        final boolean done = wrote == pending;
+        r.sendChunk(done);  // Will raise `PeerIsSlowToReadException` if the tcp send buffer is full.
+        return done;
+    }
+
+    private void sendResponse(HttpChunkedResponseSocket r, RequestState state) throws PeerIsSlowToReadException, PeerDisconnectedException {
+        boolean done;
+        do {
+            done = sendNextChunk(r, state);
```

Comment:
The pool state is closable.

The `LocalState` will logic will call its `.close()` method at the end of the request (once all `onResumeSend` calls are done) which, instead of releasing memory, returns it to the pool.

It took me a little while to figure out the lifetime of the `LocalState` and ensure all this stuff works as expected.

It's for this reason that I had to introduce a separate `free()` method to actually release the memory, called when the pool goes away.

Since we also end up with multiple prometheus processor objects (one per worker thread), I noticed that the pool wasn't getting used very well at all (via logging and debugging and testing with parallel and repeated requests). Sharing the pool across all workers fixed the issue.

Reply:
to that end, since pool is global, can we have tests that it doesn't leak

Reply:
The existing tests do that already.
I.e. they spotted leaks during development until I fixed the issue.

Reply:
Clarified over zoom.
The current `assertMemoryLeak` asserts that - once the server shuts down - there are no leaks.
There should be additional assertions that the request pool does not grow when serving multiple HTTP requests over the same TCP connection.

Reply:
Done.